Add setting to require Lens of Truth for Treasure Chest Minigame

mracsys opened the pull request:
This setting prevents players from progressing past the first chest room in the Treasure Chest minigame if they do not have the Lens of Truth. This makes brute forcing the minigame impossible. Demonstration video showing the effect with the lens effect turned on by editing memory instead of using the lens:

https://github.com/user-attachments/assets/890235ce-a172-41a7-be32-a74368aeb269

Note that while the video shows the key popping out of the chest link just opened, this error was corrected by updating the displaced code for the "right" chest to reference itself instead of the "left" chest. I can record a new video if needed.

The setting is off by default.

The setting is disabled by Treasure Chest Game small key shuffle settings other than vanilla in both the python settings list and failsafe checks in the game code.

mracsys committed:
add setting to require lens of truth for tcg minigame

mracsys committed:
use correct chest reference for popup item coords

fenhl commented:
A text box similar to what was added for adult archery in #2142 would be nice.

mracsys commented:
Would it be best to append to the Loser text, edit the salesman's text like key shuffle does, or add it somewhere else?

fenhl commented:
I think having it as a separate floating text box that appears after the get item cutscene would make sense. It matches the behavior in adult archery, is a bit more noticeable than including it in the item text, and is less weird lore-wise than having the salesman say it.

r0bd0g commented:
This looks weird b/c the small key still floats above the chest after you open it.
It seems weird that the game should overtly enforce that you must cheat at this minigame?
I feel like if skipping the lens for this being possible bothers you, just disable the check?

fenhl commented:
> This looks weird b/c the small key still floats above the chest after you open it. Can you fix the visual of which chest the key appears to have been in?

This has already been addressed as mentioned in the PR description.

> It seems weird that the game should overtly enforce that you must cheat at this minigame? (Maybe the textbox should be as if coming from navi, saying something like 'I think this game is rigged; we should come back after we've found some way to cheat!' or something.

This sounds like a good way to do it.

> Maybe let the player continue as far as red room, before enforcing the rigging, just because a rigged game that always fails after round one doesn't sound like very good rigged game design, but have the textbox regardless of how far in you fail?)

I had that idea as well, but with how it is right now, people who don't know the setting is on waste a bit less time.

> But I feel like if skipping the lens for this being possible bothers you, just disable the check?

For some context: this setting is primarily being added for racing formats like the main and multiworld tournaments as an alternative to banning 1/32 via the ruleset. Disabling the check is another approach that has been discussed and it's been decided against because of how it affects hints (the ambiguity of a woth lens is considered a good thing).

flagrama commented:
> Maybe the textbox should be as if coming from navi

Just remember Navi text is blue, not white, lol.

mracsys committed:
add tcg warning message, per-world-settings compat

mracsys commented:
I added the message and fixed an edge case for per-world settings to avoid overriding TCG keys to other worlds if the current world doesn't shuffle them and requires lens.

Demo of warning message:

https://github.com/user-attachments/assets/d21bd15b-00a0-45b2-992f-2fc656bbf471

"Eye of truth" is used so much in the vanilla game that I thought it was OK to say it that way instead of explicitly saying Lens of Truth.

If anyone wants to get fancy, it would be cool to animate Navi when buying the key from the salesman. I was thinking the same movement as waking Link up in the intro cutscene, along with an "I don't trust him. Are you sure?" second choice text box to confirm the key purchase to potentially save the player 10 rupees. Navi's codebase is beyond my ability/patience to trace down all the pieces to make that work. I don't plan to add this, but I'll merge any contributions to this PR branch if someone else wants to.

fenhl commented on data/settings_mapping.json:
```diff
@@ -348,7 +348,8 @@
             "fae_torch_count",
-            "ruto_already_f1_jabu"
+            "ruto_already_f1_jabu",
+            "tcg_requires_lens"
```
Timesavers is probably the wrong section for this. I'd suggest moving it to Misc.

mracsys replied:
done

mracsys commented:
Also on the first vs last room discussion, I can change it to any room. I wanted to be nice to the player while still making it somewhat thematic. Keeping it to the first room also lets us take a shortcut for en_changer to skip checking the current room ID before performing the switcheroo. It's not hard to add that check if needed.

mracsys committed:
move tcg_requires_lens to misc GUI section

flagrama commented:
> > Maybe the textbox should be as if coming from navi
> 
> Just remember Navi text is blue, not white, lol.

The Light Blue, not the dark blue.
![image](https://github.com/user-attachments/assets/fe5961a9-7b9b-496f-b564-79e70e0c1e3a)

mracsys committed:
use light blue for tcg warning text color

fenhl commented:
Merged in eb52eba. Still not sure why GitHub is randomly skipping marking PRs as merged.

flagrama commented:
@fenhl did you manually squash the commits when you did the merge? I don't see the commits for this PR in the Dev branch, so I think as far as git can tell this branch was never merged

fenhl commented:
Maybe that happened because I used `git checkout` to grab the [`update-presets.py`](https://github.com/fenhl/OoT-Randomizer/blob/fabef2d65c7c3585d795a3a464a4a989c12b1c23/update-presets.py) script from my branch during the merge. I'll add that file to `.git/info/exclude` instead of checking it out in the furture.
